### DISCUSSION TITLE
Plugin to ease Cross Platform projects

### DISCUSSION THREAD

**Discussion by lolgab** [com-lihaoyi/mill#2081] [Show and tell]
Cross-platform projects are complex in Mill. Much harder than in Sbt.
I've been trying many times to find a way to reduce the amount of boilerplate and duplication needed to publish a library for multiple Scala platforms ( JVM, JS, and Native ).
I came out with this plugin: https://github.com/lolgab/mill-crossplatform
It tries to ease cross-platform projects, but also to support more complex use cases. I tried with `utest`'s build and it is capable of building it.

What do you think? Do you want to try it out?

Known Drawbacks:
The plugin assumes that all platforms share the same Scala Versions. It can be not what you need in some cases, for example, when Scala Native didn't yet support Scala 3, builds needed to exclude Scala 3 only on the Scala Native projects.

**Comment by nafg** [comment]
This is great.

One question, just to be sure, I could use it for only some platforms, right? Like just JVM and JS and not native, for instance.

Also, it would be great to see an example that demonstrates a more complicated real world scenario, namely something like the following:

```mermaid
graph RL;
    Shared2 --> Shared1;
    JVM2 --> JVM1;
    JVM1 --> Shared2;
    JS2 --> JS1;
    JS1 --> Shared2;
```

Shared2 should depend on Shared1 for everything respectively, but JVM1 depends only on Shared2's JVM module and JS1 only depends on its JS module.

**Reply by lolgab** [reply]
> This is great.

Thank you! ❤️

> One question, just to be sure, I could use it for only some platforms, right? Like just JVM and JS and not native, for instance.

Yes, you can. It uses reflection to access the `jvm`, `js` and `native` `object`s but doesn't complain if they are not (unless a module with `native` wants to depend on a module without `native`, it that case it will throw an exception with an error message).

> Also, it would be great to see an example that demonstrates a more complicated real world scenario, namely something like the following:
> Shared2 should depend on Shared1 for everything respectively, but JVM1 depends only on Shared2's JVM module and JS1 only depends on its JS module.

Here you can see how a project like that is defined. It is using a SNAPSHOT release, since I fixed some things after `0.0.1`. I'm going to release version `0.0.2` very soon.

```scala
import mill._
import mill.scalalib._
import mill.scalajslib._
import mill.scalanativelib._
import $ivy.`com.github.lolgab::mill-crossplatform::0.0.2`
import com.github.lolgab.mill.crossplatform._

trait Shared extends ScalaModule {
  // ...
}

object shared1 extends Cross[Shared1]("2.13.8")
class Shared1(val crossScalaVersion: String) extends CrossPlatform {
  object jvm extends CrossPlatformCrossScalaModule with Shared {
    object test extends Tests
  }
  object js extends CrossPlatformCrossScalaModule with Shared with ScalaJSModule {
    def scalaJSVersion = "1.11.0"
    object test extends Tests
  }
}
object shared2 extends Cross[Shared2]("2.13.8")
class Shared2(val crossScalaVersion: String) extends CrossPlatform {
  override def moduleDeps = Seq(shared1(crossScalaVersion))
  object jvm extends CrossPlatformCrossScalaModule with Shared {
    object test extends Tests
  }
  object js extends CrossPlatformCrossScalaModule with Shared with ScalaJSModule {
    def scalaJSVersion = "1.11.0"
    object test extends Tests
  }
}
object jvm1 extends ScalaModule {
  def scalaVersion = "2.13.8"
  def moduleDeps = Seq(shared2("2.13.8").jvm)
  object test extends Tests
}
object jvm2 extends ScalaModule {
  def scalaVersion = "2.13.8"
  def moduleDeps = Seq(jvm1)
  object test extends Tests
}
object js1 extends ScalaJSModule {
  def scalaVersion = "2.13.8"
  def scalaJSVersion = "1.11.0"
  def moduleDeps = Seq(shared2("2.13.8").js)
  object test extends Tests
}
object js2 extends ScalaJSModule {
  def scalaVersion = "2.13.8"
  def scalaJSVersion = "1.11.0"
  def moduleDeps = Seq(js1)
  object test extends Tests
}
```

**Reply by lolgab** [reply]
[Here](https://github.com/com-lihaoyi/utest/pull/274/files) you can find another real world example with utest. I'm quite happy with the result!
